replay2/memory_remapper/memory_remapper.cc

```diff
@@ -14,4 +14,107 @@
+		const std::byte dead[2] = { std::byte(0xDE), std::byte(0xAD) };
+		for(int i = 0; i < replayAddressRange.length(); ++i) {
+			replayAddressRange.baseAddress().bytePtr()[i] = dead[i %2];
+		}
+	}
+
+	ReplayAddress MemoryRemapper::AddMapping(const MemoryObservation& observation) {
+
+		const CaptureAddress& captureAddress = observation.captureAddress();
+
+		auto replayAddressRangeAndOffset = findReplayAddressRangeAndOffset(captureAddress);
+		if(replayAddressRangeAndOffset.first != nullptr) {
+			throw AddressAlreadyMappedException();
+		}
+
+		const size_t resourceLength = observation.resourceGenerator()->length();
+		if(resourceLength == 0) {
+			throw CannotMapZeroLengthAddressRange();
+		}
+
+		std::byte* replayAllocation = new std::byte [resourceLength];
+		const ReplayAddress replayAddress(replayAllocation);
+
+		observation.resourceGenerator()->generate(replayAddress);
+
+		const CaptureAddressRange captureAddressRange(captureAddress, resourceLength);
+		const ReplayAddressRange replayAddressRange(replayAddress, resourceLength);
+
+		const auto newMapping = std::make_pair(captureAddressRange, replayAddressRange);
+		captureAddressRanges_.emplace(newMapping);
+
+		return replayAddress;
+	}
+
+	void MemoryRemapper::RemoveMapping(const CaptureAddress& captureAddress) {
+
+		auto iter = captureAddressRanges_.upper_bound(CaptureAddressRange(captureAddress, 0));
+		if(iter == captureAddressRanges_.begin()) {
+			throw AddressNotMappedException();
+		}
+
+		--iter;
+		const CaptureAddressRange& captureAddressRange = iter->first;
+		const ReplayAddressRange& replayAddressRange = iter->second;
+
+		const intptr_t offset = captureAddress.bytePtr() -captureAddressRange.baseAddress().bytePtr();
+
+		if(offset >= captureAddressRange.length()) {
+			throw AddressNotMappedException();
+		}
+
+		if(offset > 0) {
+			throw RemoveMappingOffsetAddressException();
+		}
+
+#ifndef NDEBUG
+		// In debug we'll splat all released memory with 0xDEAD before releasing it to help with debugging.
+		MarkDeadAddressRange(replayAddressRange);
+#endif
+
+		delete [] replayAddressRange.baseAddress().bytePtr();
+		captureAddressRanges_.erase(iter);
+	}
+
+	ReplayAddress MemoryRemapper::RemapCaptureAddress(const CaptureAddress& captureAddress) const {
+
+		auto replayAddressRangeAndOffset = findReplayAddressRangeAndOffset(captureAddress);
+
+		const ReplayAddressRange *replayAddressRange = replayAddressRangeAndOffset.first;
+		const intptr_t offset = replayAddressRangeAndOffset.second;
+
+		if(replayAddressRange == nullptr) {
+			throw AddressNotMappedException();
+		}
+
+		const ReplayAddress replayAddress(replayAddressRange->baseAddress().bytePtr() +offset);
+		return replayAddress;
+	}
+
+	const std::pair<const ReplayAddressRange*, intptr_t> MemoryRemapper::findReplayAddressRangeAndOffset(const CaptureAddress& captureAddress) const {
+
+		auto iter = captureAddressRanges_.upper_bound(CaptureAddressRange(captureAddress, 0));
```

Comment:
I can see that you have added comments in a middle commit but they got removed. Can you add them again?